contrib/golang/filters/http/source/golang_filter.cc

```diff
@@ -211,18 +211,24 @@ void Filter::onDestroy() {
 }
 
 // access_log is executed before the log of the stream filter
-void Filter::log(const Http::RequestHeaderMap*, const Http::ResponseHeaderMap*,
-                 const Http::ResponseTrailerMap*, const StreamInfo::StreamInfo& stream_info,
-                 Envoy::AccessLog::AccessLogType) {
+void Filter::log(const Http::RequestHeaderMap* headers, const Http::ResponseHeaderMap*,
+                 const Http::ResponseTrailerMap*, const StreamInfo::StreamInfo&,
+                 Envoy::AccessLog::AccessLogType type) {
   // `log` may be called multiple times with different log type
-  if (!stream_info.requestComplete().has_value()) {
-    return;
+  auto& state = getProcessorState();
+
+  if (req_ == nullptr) {
+    // log called by AccessLogDownstreamStart will happen before doHeaders
+    initRequest(state);
+
+    request_headers_ = static_cast<Http::RequestOrResponseHeaderMap*>(
+        const_cast<Http::RequestHeaderMap*>(headers));
   }
 
-  // TODO: support mid-request logging
-  auto& state = getProcessorState();
-  state.log();
-  dynamic_lib_->envoyGoFilterOnHttpLog(req_);
+  state.enterLog();
+  req_->phase = static_cast<int>(state.phase());
+  dynamic_lib_->envoyGoFilterOnHttpLog(req_, int(type));
```

Comment:
I think we'd better only call it when the type is `AccessLogDownstreamStart`, `AccessLogDownstreamPeriodic`, or `AccessLogDownstreamEnd`, which is supported now.
I'm a little worried about the performance, of introducing the cgo calls when they are useless.

Eventually, we'd better allow people to configure which will be enabled on the envoy side. for now, only these three types, I think it might be acceptable.

Reply:
Updated.